### DISCUSSION TITLE
Support Kotlin + Swift in Turbo Module with backward compat

### DISCUSSION THREAD

**Discussion by vicary** [callstack/react-native-builder-bob#351] [Ideas]
We have some hardware SDKs (thermal printers) written in Kotlin and Swift, we would like to use `create-react-native-library` to create a React Native wrapper for both of them.

Currently for the *Turbo Module with backward compat* option it only allows Kotlin + Objective-C, is it possible to create Kotlin + Swift template?

**Comment by atlj** [comment]
> We have some hardware SDKs (thermal printers) written in Kotlin and Swift, we would like to use `create-react-native-library` to create a React Native wrapper for both of them.
> 
> 
> 
> Currently for the *Turbo Module with backward compat* option it only allows Kotlin + Objective-C, is it possible to create Kotlin + Swift template?

Currently there is no Swift template for TurboModules. Here how you can work with Swift on TurboModules for the time being:

1. Open the module on Xcode ( not the example app )
2. Create a Swift file and make sure to create the brigding header.
3. Add @objc public to your classes and methods you want to use
4. Go to your example app's project on Xcode
5. Go to the .mm file
6. Import the swift headers (they are named like the following: (your podspec name)-Swift.h
7. Now you can import the swift methods and classes from objective c++

You can also take a look at React Native Community's sample https://github.com/react-native-community/RNNewArchitectureLibraries/tree/feat/turbomodule-swift

**Reply by vicary** [reply]
Looking at any one sources really doesn't give enough knowledge over these topics:
1. XCode configurations
2. Swift
3. Bridging Headers
4. Objective C++ (I combined the `.h` and `.mm` from your objc template into one single `.m` for Swift)
5. How to properly place macros for both oldarch and newarch

After a few days of crawling through different sources, I think I did the job and made [react-native-citizen-escposprinter](https://www.npmjs.com/package/react-native-citizen-escposprinter) works in all four environments.

**Reply by antonyjoemathew** [reply]
@atlj @vicary I'm working on creating a React wrapper for my Swift SDK, and I've followed the same approach. However, I'm facing an issue where I cannot import my Swift module in the Appdelegate.mm file. Are there additional steps or considerations I should take to address this issue?

**Reply by vicary** [reply]
My memory really faded after almost a year, since my package is working properly since then.

The first thing comes into my mind is that for both newarch and "oldarch" to work, you have to ditch .h and .mm and use a single .m to bridge your Swift module into objc

**Reply by antonyjoemathew** [reply]
Thank you @vicary  for your quick reply:)

**Reply by gladiuscode** [reply]
@antonyjoemathew Hi!

Where you able to run your Swift code inside the AppDelegate.mm file?
I'm trying right know but I'm stuck. Xcode isn't generating my Pod-Name-Swift.h so my app can't compile.

I've got the following files: Name.m, Name.swift and Pod-Name-Bridging-Header.h.
My swift class is public and has @objc annotation.

Any help would be gladly appreciated :)

**Comment by JeanSadatAt** [comment]


**Reply by vicary** [reply]
What is this? Some kind of ChatGPT spamming for online presence?

**Reply by vicary** [reply]
@atlj tagging you for you are a maintainer. Not sure what to do yet, just raising awareness.

GitHub doesn't seem to allow reporting discussion comments, but this user is suspicious enough. With the recent hype on ChatGPT, this may be the beginning of something really annoying.

**Reply by MateWW** [reply]
@vicary I've deleted it as nonsense truly generated by chat GPT. 
Thanks for reporting!

**Reply by vicary** [reply]
@MateWW love the swift action *(pun intended)*

**Reply by atlj** [reply]
Deleting comments generated by ChatGPT. We truly became Stack Overflow 😅

**Reply by vicary** [reply]
These junks will inevitably be fed to future iterations such as GPT-4, the cycle has begun.

**Comment by troZee** [comment]
Probably this issue is related to https://github.com/callstack/react-native-builder-bob/discussions/342

**Comment by antonyjoemathew** [comment]
Hi,

I've resolved the problem by creating Objective-C files for the bridging,
and from there, I called the functions in the Swift files

On Thu, 16 May 2024 at 03:25, gladius ***@***.***> wrote:

> @antonyjoemathew <https://github.com/antonyjoemathew> Hi!
>
> Where you able to run your Swift code inside the AppDelegate.mm file?
> I'm trying right know but I'm stuck. Xcode isn't generating my
> Pod-Name-Swift.h so my app can't compile.
>
> I've got the following files: Name.m, Name.swift and
> Pod-Name-Bridging-Header.h.
> My swift class is public and has @objc <https://github.com/objc>
> annotation.
>
> Any help would be gladly appreciated :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/callstack/react-native-builder-bob/discussions/351#discussioncomment-9451014>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACESDVQAAWEJI3PWCIMFKSDZCPKWVAVCNFSM6AAAAAAUB24DYWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TINJRGAYTI>
> .
> You are receiving this because you were mentioned.Message ID:
> <callstack/react-native-builder-bob/repo-discussions/351/comments/9451014@
> github.com>
>


-- 
Regards
Antony Joe Mathew
+971555167780
+919496371536

**Reply by vicary** [reply]
Would you please share some high level steps so people can follow in the future?

**Reply by gladiuscode** [reply]
@vicary I've found a way.
I'll document the entire process in the this [thread](https://github.com/callstack/react-native-builder-bob/discussions/554) that I've opened specifically for the library I'm developing.

Thanks you all!

**Comment by snowinszu** [comment]
Hello, any news? Will Kotlin + Swift in Turbo Module be supported soon?
